core/sdk-core/src/main/java/software/amazon/awssdk/core/progress/listener/ProgressListener.java

```diff
@@ -0,0 +1,395 @@
+ *     <li>Be mindful that {@link #requestBytesSent(Context.RequestBytesSent)} or
+ *     {@link #responseBytesReceived(Context.ResponseBytesReceived)}
+ *     may be called extremely often for large payloads
+ *     (subject to I/O buffer sizes). Be careful in implementing expensive operations as a side effect. Consider rate-limiting
+ *     your side effect operations, if needed.</li>
+ *     <li>{@link ProgressListener}s may be invoked by different threads. If your {@link ProgressListener} is stateful,
+ *     ensure that it is also thread-safe.</li>
+ *     <li>{@link ProgressListener}s are not intended to be used for control flow, and therefore your implementation
+ *     should not <i>throw</i>. Any thrown exceptions will be suppressed and logged as an error.</li>
+ * </ol>
+ * <p>
+ * A classical use case of {@link ProgressListener} is to create a progress bar to monitor an ongoing transfer's progress.
+ * Refer to the implementation of {@link LoggingProgressListener} for a basic example, or test it in your application by providing
+ * the listener as part of your {@link SdkHttpRequest}. E.g.,
+ * <pre>{@code
+ * AmazonS3Client s3Client = new AmazonS3Client(awsCredentials);
+ * PutObjectRequest putObjectRequest = PutObjectRequest.builder()
+ *                                        .bucket("bucket")
+ *                                        .key("key")
+ *                                        .overrideConfiguration(o -> o.addProgressListener(LoggingTransferListener.create())
+ *                                        .build();
+ * s3Client.putObject(putObjectRequest);
+ * }</pre>
+ * And then a successful transfer may output something similar to:
+ * <pre>
+ * Request initiated...
+ * |                    | 0.0%
+ * |==                  | 12.5%
+ * |=====               | 25.0%
+ * |=======             | 37.5%
+ * |==========          | 50.0%
+ * |============        | 62.5%
+ * |===============     | 75.0%
+ * |=================   | 87.5%
+ * |====================| 100.0%
+ * Request execution successful!
+ * </pre>
+ */
+@SdkPublicApi
+public interface ProgressListener {
+
+    /**
+     * This method is called right after a request object is marshalled and ready to be sent to the service
+     * <p>
+     * Available context attributes:
+     * <ol>
+     *     <li>{@link Context.RequestPrepared#request()}</li>
+     *     <li>{@link Context.RequestPrepared#progressSnapshot()}</li>
+     * </ol>
+     */
+    default void requestPrepared(Context.RequestPrepared context) {
+    }
+
+    /**
+     * This method is called after the request transaction is initiated, i.e. request header is sent to the service
+     * <p>
+     * Available context attributes:
+     * <ol>
+     *     <li>{@link Context.RequestHeaderSent#request()}</li>
+     *     <li>{@link Context.RequestHeaderSent#progressSnapshot()}</li>
+     * </ol>
+     */
+    default void requestHeaderSent(Context.RequestHeaderSent context) {
```

Comment:
Can we also add to the Javadoc of each what methods might be called next? E.g. after this will always be either requestBytesSent, responseHeaderReceived, or attemptFailed?

Reply:
Done